Refactor IOBuffer code

jakobnissen opened the pull request:
I've been a little frustrated with the IOBuffer code. It contains a whole bunch of implicit invariants, and is poorly commented. It also has several bugs that ultimately stems from the code being unclear about its own assumptions.

This is a refactoring of IOBuffer. The primary goals are:
* Comment the code more heavily
* Test the code more thoroughly

The secondary goals are
* Fix a few outstanding bugs
* Add some minor performance improvements

This is a purely internal refactoring with be no change in behaviour of `IOBuffer`, except straight up bugfixes. However, note that previous code may have relied on buggy behaviour. Fixing bugs may therefore cause breakage.

## Current changes
### **BEHAVIOUR CHANGES**
* The following code used to not throw an error, but now does: `IOBuffer(b"abc"; maxsize=2)`. I consider this a bugfix. It should not be possible to construct an IOBuffer with a buffersize larger than `maxsize`.
* It used to be possible to write to indices higher than `maxindex`, which could trigger a bug causing data loss. The bug has been fixed, but as a result, some IOBuffers may reach full capacity faster (really: reach it at the correct point), changing writing behaviour. 

### Bugfixes
* Do not corrupt data on `copyline` on a non-appending buffer
* Respect `maxsize` even after `take!` (fix #57549)
* Fix bug when copying from an appending iobuffer to itself
* Fix bug where re-allocating the buffer may cause it to shrink, discarding data.
* Fix bug where `truncate` may throw a wrong BoundsError
* Fix a bug where truncating a buffer may not correctly removed mark at position that has been deleted
* Fix a bug where initializing an IOBuffer without an explicit buffer and with `truncate=false` makes it contain the full buffer
* Current behaviour for `reset` and `position` did not work for `PipeBuffer`. Fix that

### Changes to brittle code
* Removed some tests that explicitly tested internal code and internal behaviour. Some of that behaviour has changed.
* Changed some internal `PipeBuffer` behaviour, which did not respect writable IOBuffer's ownership of their buffer and therefore failed spuriously

### Performance improvements
* Writing to dense IOBuffer now uses memmove and is up to 10x faster for long writes.
* Minor optimisations (about ten percent) for writing to IOBuffers in general 

Closes #57549

jakobnissen committed:
WIP: Comment IOBuffer code

jakobnissen committed:
WIP: Add some more tests

vtjnash commented on base/iobuffer.jl:
```diff
@@ -2,19 +2,75 @@
+
+    # Data is read/written from/to ptr, except in situations where append is true, in which case
+    # data is still read from ptr, but written to size+1.
+    # This value is alwaus in offset+1 : size+1
```
This is actually a bug in the current implementation. The ptr is not supposed to be restricted to be inside the file size, and any calls to write that happen should then automatically zero-pad the file from size:ptr (https://pubs.opengroup.org/onlinepubs/009695399/functions/lseek.html)

jakobnissen replied:
Hmm how does that work? Would the user have to `seek` first to beyond `size`? That doesn't correspond to how seek normally works in Julia: Normally, you can't seek beyond the filesize. Should that be allowed? Then what happens if the user seeks beyond `lastindex(buffer.data)`?

jakobnissen replied:
How is it supposed to work with `IOStream`? What file should the following script create?
```julia
f = open("/tmp/foo", "w+")
write(f, "abcd")
flush(f)
seek(f, 10)
write(f, 'a')
flush(f)
```

jakobnissen commented on base/iobuffer.jl:
```diff
@@ -329,6 +404,7 @@ function _resize!(io::GenericIOBuffer, sz::Int)
     a = io.data
     offset = io.offset
     if applicable(resize!, a, sz)
+        # TODO: This is buggy: The buffer should never touch data before the offset.
```
That is good to know - adding the comments is precisely to learn what is part of the contract! Okay, so the `IOBuffer` takes ownership of its underlying array - and, in the case of being passed a `Vector`, it then transitively takes control of the underlying `Memory`.

jakobnissen committed:
Update comments after feedback

jakobnissen committed:
Remove offset field from GenericIOBuffer

This field was added in d7b9ac8 when the buffer's underlying data was changed
to Memory. However, I think it's not a good tradeoff. The benefit of the offset
field is that, when a buffer is constructed from a Vector which has previously
been pushfirst!'ed to, no data needs to be copied into the vector.
However, this is probably rare.
The downside is that it makes the implementation more complex, the struct larger,
and adds extra overhead to all operations on the IO.

In this commit, we instead copy the data internally in the buffer if constructed
from a Vector with a non-zero memory offset.

jakobnissen committed:
Refactor resizing

* Remove tests that tests internal, unexported functions and messes with the
  internal fields of IOBuffer.
* Do not overallocate in `_resize`, rather, overallocate in the callers.
  This makes responsibility over determining overallocation more clear
* Refactor ensureroom, changing the logic for when the data is shifted in the
  buffer.

jakobnissen committed:
Update comments to mention mark is zero-indexed

jakobnissen committed:
Handle maxsize and sizehint better on IOBuffer construction
* Correctly throw ArgumentError if maxsize is < 0
* Do not preallocate the maximal buffersize, this is what sizehint is for.

jakobnissen committed:
Allow read and peek from non-dense IOBuffer

This allows e.g. reading from an IOBuffer backed by a UnitRange or whatever.

jakobnissen committed:
Also reinit used-up non-Memory buffer after take!

jakobnissen committed:
Use memcpy to write to IOBuffer when applicable

* For non-memory backed buffers, do not use invalid inbounds.
* For memory-backed buffers, use unsafe_copyto!, which is much faster

jakobnissen committed:
Simplify and speed-up occursin implementation

Use the existing implementation of in directly with the buffer. This should
already be fast

jakobnissen committed:
Bugfix: Buffer behaviour in unsafe_read

This fixes a bug introduced somewhere earlier in this chain of commits. However,
the previous behaviour wasn't great either.
In the changed code, correctness relied on the buffer having very specific
resizing behaviour, i.e. a mercurial implementation detail. The current solution
is more robust. However:
1. This workaround copies the data one extra time, which is not great
2. It's possible that the bug surfaced here due to another bug in the resizing
   behaviour. However, if so, I haven't been able to find it.

jakobnissen committed:
Add more tests

jakobnissen committed:
Bugfix: Fix copyline for non-appending IOBuffer

jakobnissen commented:
Let's stop the PR here - tests pass, the most pressing issues has been addressed, and it's already a fairly large (too large?) PR. I'll make a followup PR with some more tests after this is merged and I get code coverage.

jakobnissen commented:
Adding @nhz2 because you've been a thorough reviewer on previous IO related PRs. Feel free to unassign yourself.

jakobnissen committed:
Minor optimisation of write(::IOBuffer, mem)

* Outline slow path in ensureoom.
* Do not uses memcpy when writing short arrays

jakobnissen committed:
Enforce invariants in IOBuffer constructors

jakobnissen committed:
Tweak compacting heuristics

jakobnissen committed:
Bugfix: Copying to self with appending buffer

KristofferC commented on base/iobuffer.jl:
```diff
@@ -1,26 +1,101 @@
+
+# Internal trait object used to access unsafe constructors.
+struct Unsafe end
+const unsafe = Unsafe()
```
These names feel a bit too generic to put into the `Base` namespace.

jakobnissen replied:
Changed the name to `UnsafeMethod` and `unsafe_method`. They can be called something arbitrary, but I do think it's nice that they're available for all internal, unsafe methods.

jakobnissen committed:
Rename Unsafe to UnsafeMethod

jakobnissen committed:
Fix errors during build

This is due to some sketchy behaviour by some other code which relies on IOBuffer
internals, but shouldn't.

jakobnissen committed:
Fix typo in test

jakobnissen committed:
Skip initializing IOBuffer data

Before Julia 1.11, observing uninit bitstypes was UB. Now, it's no longer UB,
so no need.

nhz2 commented on base/iobuffer.jl:
```diff
@@ -29,14 +104,20 @@ function GenericIOBuffer(data::T, readable::Bool, writable::Bool, seekable::Bool
     ref = data.ref
-    buf = GenericIOBuffer(ref.mem, readable, writable, seekable, append, maxsize)
+    mem = ref.mem
+    len = length(data)
```
If `len < maxsize < length(mem)` could this trip the `"maxsize must not be smaller than data length"` error in an unexpected way?

jakobnissen replied:
Good catch. I've addressed this by setting `maxsize` to `length(mem)` in that case. That's not ideal - since if the user `take!`s the buffer, and then re-initializes it, it may grow to larger than expected.
However, I can't think of a better way to handle it. Alternatively, we could allow `maxsize` to be smaller than the buffer length in that particular case.

nhz2 commented on base/iobuffer.jl:
```diff
@@ -131,17 +210,26 @@ function IOBuffer(;
+    # as the default storage, we can turn the result into a string without copying.
+    data = fill!(StringMemory(size), 0)
+    buf = GenericIOBuffer{Memory{UInt8}}(unsafe_method, data, flags.read, flags.write, true, flags.append, mz)
+    if flags.truncate
+        buf.size = 0
+    end
```
Since there is no initial data, I think it makes sense to always set `size` to zero.

If `size` is zero, you can probably remove the `data = fill!(StringMemory(size), 0)`.

Otherwise, passing `truncate=false` exposes the uninitialized memory for reading.

```julia-repl
julia> read(IOBuffer(;maxsize=34, truncate=false))
34-element Vector{UInt8}:
 0x00
 0x00
 0x00
 0x00

```

nhz2 commented on base/iobuffer.jl:
```diff
@@ -221,7 +308,45 @@ function unsafe_read!(dest::Ptr{UInt8}, src::DenseBytes, so::Integer, nbytes::UI
+    for i in 0:sizeof(x)-1
+        x |= typeof(x)(data[from + i]) << (8 * i)
+    end
+    reinterpret(T, ltoh(x))
```
I don't think `ltoh` should be used here, as `read` and `write` are documented to not convert endianness.

jakobnissen replied:
What does it mean that it doesn't convert endianness? If the system is big-endian, then the code in `load_from_array` is just wrong, no? I mean, then it will load it as little-endian.

nhz2 replied:
Yes you are right.

nhz2 commented on base/iobuffer.jl:
```diff
@@ -235,26 +360,20 @@ function peek(from::GenericIOBuffer, T::Union{Type{Int16},Type{UInt16},Type{Int3
 end
 
-function read_sub(from::GenericIOBuffer, a::AbstractArray{T}, offs, nel) where T
+function read_sub(from::GenericIOBuffer, a::MutableDenseArrayType{T}, offs, nel) where T
```
This whole function should be removed. It is only used in `readbytes!` and there it should be replaced with `copyto!` like what I do in https://github.com/JuliaIO/InputBuffers.jl/blob/da9b2315f11cd041a247bb2a088d6e1f28ccd9a4/src/InputBuffers.jl#L79-L88

nhz2 commented on base/iobuffer.jl:
```diff
@@ -1,26 +1,101 @@
+    # Data at the marked location or before for non-seekable buffers can be deleted.
+    # The mark is zero-indexed. If it is -1, the mark is not set.
+    # The purpose of the mark is to reset the stream to a given position using reset.
+    # This value is always in -1 : size-1
```
```suggestion
    # This value is always in -1 : size
```

oscardssmith commented on base/iobuffer.jl:
```diff
@@ -1,26 +1,101 @@
+    # Data is read/written from/to ptr, except in situations where append is true, in which case
+    # data is still read from ptr, but written to size+1.
+    # This value is always in 1 : size+1
+    ptr::Int
```
I wonder if we could make this a `MemoryRef` If possible, doing so might save a few cycles.

jakobnissen replied:
That can't be done, because a MemoryRef can't point to one past the end, which ptr needs to do to signal EOF

vtjnash replied:
Oscar seems probably right that we could replace the Memory+offset fields with a single MemoryRef field, since that may express almost the same concept

vtjnash commented:
The removal of offset is a breaking change and performance regression. That probably shouldn't be mixed in with a PR that claims to be just bug fixes and refactoring for clarity

jakobnissen commented:
I would argue against it being a performance regression, but surely, it's not a breaking change. How is it breaking to remove a non-documented field of a struct? One which was added only in the current release? Which documented behaviour does that break?

jakobnissen committed:
Reinstate offset field

Also test the behaviour that depends on this field

jakobnissen committed:
Address some reviewer comments

jakobnissen committed:
Try to make inferrible for trimming CI

jakobnissen committed:
Minor change in comments

jakobnissen committed:
DECISION: Do not fill buffer past maxsize

caleb-allen commented on base/iobuffer.jl:
```diff
@@ -283,27 +410,41 @@ read(from::GenericIOBuffer, ::Type{Ptr{T}}) where {T} = convert(Ptr{T}, read(fro
+# Position is zero-indexed, but ptr is one-indexed, hence the -1
+# TODO: Document that position for an unseekable stream is invalid, or
+# make it error
 position(io::GenericIOBuffer) = io.ptr - io.offset - 1
```
My vote here is for an error

nhz2 replied:
`mark` returns the "marked position" of the stream, so erroring will also make `mark` error.

jakobnissen committed:
More efficient copy of IOBuffer

Only copy the used bytes, not unreachable bytes.

jakobnissen commented:
Thanks for the review. All comments has now been addressed (except the one about endianness which I don't understand)

jakobnissen commented:
The build failure (segfault) on 32-bit linux looks unrelated.

nsajko commented on base/iobuffer.jl:
```diff
@@ -1,26 +1,121 @@
+    mark::Int
+
+    # Unsafe constructor which does not do any checking
+    function GenericIOBuffer{T}(
+            ::UnsafeMethod,
```
This design, where some `Method`s are intended as private, does not lead to a good UX. Here, the `methods` function doesn't know that the internal constructor method is supposed to be private, so it's shown together with the other methods:

```julia-repl
julia> methods(IOBuffer)
# 6 methods for type constructor:
 [1] IOBuffer(s::SubString{String})
     @ strings/io.jl:309
 [2] IOBuffer(str::String)
     @ strings/io.jl:308
 [3] IOBuffer(; read, write, append, truncate, maxsize, sizehint)
     @ iobuffer.jl:231
 [4] IOBuffer(data::AbstractVector{UInt8}; read, write, append, truncate, maxsize, sizehint)
     @ iobuffer.jl:211
 [5] Base.GenericIOBuffer{T}(data::T, readable::Bool, writable::Bool, seekable::Bool, append::Bool, maxsize::Integer) where T<:AbstractVector{UInt8}
     @ iobuffer.jl:104
 [6] Base.GenericIOBuffer{T}(::Base.UnsafeMethod, data::T, readable::Bool, writable::Bool, seekable::Bool, append::Bool, maxsize::Int64) where T<:AbstractVector{UInt8}
     @ iobuffer.jl:90
```

At best that's a distraction/noise. What's worse is, the fact that the "private" method is shown together with the other ones could encourage inexperienced or unruly users to use the private method.

Instead of the design with the private/unsafe method, I propose having the inner constructor be a private function. Say, `_new_generic_io_buffer`:

```suggestion
    global function _new_generic_io_buffer(
            ::Type{T},
```

nsajko commented on base/iobuffer.jl:
```diff
@@ -110,15 +215,13 @@ function IOBuffer(
     end
     flags = open_flags(read=read, write=write, append=append, truncate=truncate)
-    buf = GenericIOBuffer(data, flags.read, flags.write, true, flags.append, Int(maxsize))
+    buf = GenericIOBuffer(data, flags.read, flags.write, true, flags.append, maxsize)
```
You seem to be aware of what I wrote, looking at the rest of the code, so I'm not sure why you removed to `Int(...)` call here?

nsajko commented on base/iobuffer.jl:
```diff
@@ -110,15 +215,13 @@ function IOBuffer(
-    if maxsize < 0
-        throw(ArgumentError("negative maxsize"))
-    end
+        sizehint::Union{Integer,Nothing}=nothing,
+    )
```
Probably it's better to keep the diff smaller by avoiding unrelated stylistic changes?

```suggestion
        sizehint::Union{Integer,Nothing}=nothing)
```

jakobnissen replied:
Thanks for the review - this and the other comment of you have been addressed as you suggested

jakobnissen committed:
Address reviewer comments

* Remove unsafe trait objects, to avoid adding private methods to public
  functions
* Remove a superfluous style change

nsajko commented on base/iobuffer.jl:
```diff
@@ -164,15 +268,43 @@ PipeBuffer(maxsize::Integer) = (x = PipeBuffer(StringMemory(maxsize), maxsize =
+    else
+        # When the buffer is just readable, they can share the same data, so we just make
+        # a shallow copy of the IOBuffer struct.
+        # Use unsafe method because we want to allow b.maxsize to be larger than data, in case that
```
The wording has to be updated now, maybe like this:

```suggestion
        # Use internal constructor because we want to allow b.maxsize to be larger than data, in case that
```

jakobnissen replied:
Done

jakobnissen committed:
Fixup

jakobnissen commented:
Bump. This is done now - a few people have reviewed it, but none have approved. It would be nice to get this in early in 1.13's cycle, so any bugs are detected early. Once this PR has landed, and #57668 has been fixed, I can make a followup PR to get test coverage closer to 100%

vtjnash commented on base/stream.jl:
```diff
@@ -617,7 +617,7 @@ end
     ensureroom(buffer, Int(recommended_size))
     ptr = buffer.append ? buffer.size + 1 : buffer.ptr
-    nb = min(length(buffer.data)-buffer.offset, buffer.maxsize) + buffer.offset - ptr + 1
+    nb = min(length(buffer.data), buffer.maxsize) - ptr + 1
```
This is now only lets writing up to `maxsize-offset` bytes, which doesn't seem intended

jakobnissen replied:
It seems to be correct, yes.
1. We can't write pas `maxsize` when `offset` is zero, because if we do, then resizing or replacing the buffer (in an unrelated operation) may cause data loss, since the buffer is now smaller.
2. The call to `ensureroom` ensures there is at least `recommended_size` bytes available. If offset is not zero, the offset is zeroed (moving the content of the data).

Therefore, the pointer and length returned by `alloc_request` returns the correct maximum number of bytes that can be written, at least `recommended_size`, and only shifts the data inside the buffer if necessary to achieve `recommended_size`.

vtjnash replied:
maxsize - ptr is a negative number (equal to `-offset`) when you first created the IOBuffer with maxsize bytes available, so that can't be right

vtjnash replied:
The value `length(buffer.data)` includes offset, but buffer.maxsize does not, so these two input values to `min` do not have the same units

vtjnash replied:
Yes, for the case now using `get_offset`

vtjnash commented on base/stream.jl:
```diff
@@ -942,19 +942,20 @@ function readbytes!(s::LibuvStream, a::Vector{UInt8}, nb::Int)
-        initsize = length(a)
-        newbuf = PipeBuffer(a, maxsize=nb)
-        newbuf.size = newbuf.offset # reset the write pointer to the beginning
+        # TODO: Here, we don't read directly into `a`, but read to
+        # a new buffer, and then copy to `a`
+        newbuf = PipeBuffer(; maxsize=nb)
```
This seems to be a regression?

jakobnissen replied:
This here is the same as the next three. `IOBuffer` owns the underlying data once it has been written to. However, the previous code here attempted to save an allocation by wrapping an existing vector in an `IOBuffer`, and then writing to the vector by writing to the `IOBuffer`, before using the vector.
This is invalid, because the `IOBuffer` owns the vector.
To see why it's an issue, it means that this code relies on undocumented and possibly changing details about how the `IOBuffer` reallocates the buffer. For example, if we ever decide to reallocate the memory of `IOBuffers` pre-emptively if enough data is written to them, this, seemingly unrelated code will break.

vtjnash replied:
This was the original purpose of IOBuffer though, not the other way around. The original intent was supporting zero-copy IO specifically for this file, while it just happens to also be a fairly usable IO abstraction also.

> This is invalid, because the IOBuffer owns the vector.

That is true, then we destroy the IOBuffer at the end (_unsafe_take!) and rebuild the vector (_take!) to transfer ownership back to the Vector so that at any given time there is only one owner of the buffer

vtjnash commented on base/stream.jl:
```diff
@@ -991,12 +992,17 @@ function unsafe_read(s::LibuvStream, p::Ptr{UInt8}, nb::UInt)
     else
-        newbuf = PipeBuffer(unsafe_wrap(Array, p, nb), maxsize=Int(nb))
-        newbuf.size = newbuf.offset # reset the write pointer to the beginning
+        # TODO: This isn't great. Ideally, we would want to write directly to the
+        # pointer, instead of doing what we do here, which is to write to a new
+        # IOBuffer, and then copy the content of the IOBuffer into the pointer.
+        newbuf = PipeBuffer(; maxsize=Int(nb))
```
Why this regression also?

vtjnash commented on base/stream.jl:
```diff
@@ -1599,19 +1605,20 @@ function readbytes!(s::BufferStream, a::Vector{UInt8}, nb::Int)
-            initsize = length(a)
-            newbuf = PipeBuffer(a, maxsize=nb)
-            newbuf.size = newbuf.offset # reset the write pointer to the beginning
+            # TODO: Copy data directly from s into a. Instead, here,
+            # we use a PipeBuffer as indirect IO.
+            newbuf = PipeBuffer(; maxsize=nb)
```
And this regression?

vtjnash commented on base/iobuffer.jl:
```diff
@@ -581,31 +831,27 @@ end
-        b == delim && return true
-    end
-    return false
+    ptr = buf.ptr
+    return in(delim, view(buf.data, ptr:ptr + bytesavailable(buf)-1))
```
Seems a slightly awkward way to compute this, since buf.size is already the pointer to the end of the file, while bytesavailable is a computation based on that which this then had to undo
```suggestion
    return in(delim, view(buf.data, buf.ptr:buf.size))
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -536,38 +739,85 @@ _unsafe_take!(io::IOBuffer) =
+    append = to.append
+    ptr = append ? size+1 : to.ptr
+    data = to.data
+    to_write = min(nb % Int, min(Int(length(data))::Int, to.maxsize) - ptr + 1)
```
Comparison should be done as unsigned to avoid large nb becoming a negative value, since it is the only number here that we don't know has a reasonable value
```suggestion
    to_write = min(nb, (min(Int(length(data))::Int, to.maxsize) - ptr + 1) % UInt) % Int
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -536,38 +739,85 @@ _unsafe_take!(io::IOBuffer) =
+            iszero(to_write) && return 0
+            GC.@preserve from unsafe_copyto!(data, size + 1, data, from.ptr, to_write)
+            from.size = size + to_write
+            from.ptr += to_write
```
even if the write failed, in theory we still read that data
```suggestion
            from.ptr = from.size + 1
```

(I'd also support just making this whole case an error, since it should really be an infinite loop per the generic definition of reading until eof)

jakobnissen replied:
What you suggest here is inconsistent with how `write(::IOBuffer, ::IOBuffer)` currently works when it's to distinct buffers. Consider:
```julia
julia> buf = IOBuffer(;maxsize=3);
       buf2 = IOBuffer(collect(b"abcdefgh"));
       write(buf, buf2)
       println(String(read(buf2)))
defgh
```
If I understand correctly, you suggest that in this example, `buf2` should have been read until completion, and should now be empty?

I see that the existing behaviour is different from the generic definition:
```julia
julia> buf = IOBuffer(;maxsize=3);
       buf2 = IOBuffer(collect(b"abcdefgh"));
       @invoke write(buf::IO, buf2::IO)
       eof(buf2)
true
```
So I'll change the behaviour in the following way:
* Error when `to === from`, as you suggest
* Make sure that, when `to !== from`, all of `from` is read, even if `to` is full.

vtjnash replied:
Yes, that is correct, I guess I forgot to update the read pointer on `from` when `from !== to`. Good catch

vtjnash commented on base/iobuffer.jl:
```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
+        # Only shift data if:
+        if (
+                # It will prevent us from having to resize buffer, or
+                to_delete >= nshort % Int ||
```
I realize this is sort of what we did before (with a threshold value of 4096 instead of 1 though), but this is not actually a good idea to have here since it can make writing to a PipeBuffer `O(n^2)` in the worst case. We should instead resize anytime that we would recover less than 32 bytes or less than 1/8th of the buffer, whichever is larger

jakobnissen replied:
Did as you suggested, except I also added a check to shift in case `io.maxsize` is low enough that resizing wouldn't free enough space.

vtjnash commented on base/iobuffer.jl:
```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
+            )
+            copyto!(data, 1, data, to_delete + 1, size - to_delete)
+            io.ptr = ptr - to_delete
+            io.mark = max(-1, mark - to_delete)
```
Reading the documentation for `mark`, this field value is not permitted to change here (we seem to have handled this wrong before too). I think you need to change offset instead, to essentially virtualize the position:
```suggestion
            io.offset -= to_delete
```

(although having a negative offset is slightly worrying, I think most computations actually should handle it fine already, but we should try to double-check that)

jakobnissen replied:
I don't think a negative offset would be a correct solution to this problem. The offset already means something specific: The leading number of noncoding bytes in the data. Adding a separate meaning will lead to bugs.
However, perhaps the `offset` field can be stored in an encoded format as a `UInt`, where the top bit signals if the offset means "number of trailing unused bytes in the buffer", or it means "number of previously discarded, coding bytes after compacting". Only one of these will be needed at any one time, the offset will always be zero after compacting has happened.

jakobnissen replied:
Maybe that's precisely what you meant - anyway, will give it a shot

vtjnash replied:
That seems close to what I meant, though I think the math usually works out such that those two different encodings definitions would always have the same numeric value: since we never access the data in the buffer before mark and that mark is relative to offset. This computationally is basically an OffsetArray, where accessing any data < mark is defined to throw an exception, so we don't actually need to keep that data, but may need to adjust the firstindex (aka offset) such that the computation of the mark location remains stable even if as it shifts the underlying indexes

jakobnissen replied:
This has been addressed in (one of) the latest commits: https://github.com/JuliaLang/julia/pull/57570/commits/819d7506b51d385e534d043042567750687cf621. Might not be exactly what you envisioned, but it works and should be efficient.

vtjnash commented on base/iobuffer.jl:
```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
     return io
 end
 
+# TODO: These errors cannot be converted to LazyString, but it's wasteful to interpolate them here.
```
It is generally best not to have comments saying something cannot be done, since they have a tendency to become stale, and then prevent someone else from considering doing the very thing the comment wanted to do
```suggestion
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
+function _resize!(io::GenericIOBuffer, new_size::Int)
+    old_data = io.data
+    if applicable(resize!, old_data, new_size)
+        resize!(old_data, new_size)
```
This is already defined to call `overallocation`, so now we're calling it twice (doubling the wasted space) if anyone was using an IOBuffer wrapping something mutable. Could you sink the call to `overallocation` into this function, and instead add another (exact::Bool) that controls whether it gets called? That may also avoid needing to add the requirement to call `offset = 0` first, which can result in redundant and unnecessary data copies in some cases in the old buffer, when the data needed to be moved to the new buffer anyways.

vtjnash commented on base/iobuffer.jl:
```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
     return io
 end
 
+# TODO: Should check for seekable and error if not
```
seekend is defined for all streams (forward skips are always allowed)
```suggestion
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -283,27 +436,41 @@ read(from::GenericIOBuffer, ::Type{Ptr{T}}) where {T} = convert(Ptr{T}, read(fro
-        io.ptr += min(n, n_max)
+        # Don't use seek in order to allow a non-seekable IO to still skip bytes.
+        # Handle overflow
+        io.ptr = min(io.size + 1, clamp(widen(io.ptr) + widen(n), Int))
```
nit: avoid `clamp+widen` by writing out the overflow check?
```suggestion
        io.ptr = min(io.size + 1, typemax(Int) - n > io.ptr ? typemax(Int) : io.ptr + n)
```
or even just expand out the whole expression:
```suggestion
        maxptr = io.size + 1
        io.ptr = n > maxptr || maxptr - n > io.ptr ? maxptr : io.ptr + n
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -283,27 +436,41 @@ read(from::GenericIOBuffer, ::Type{Ptr{T}}) where {T} = convert(Ptr{T}, read(fro
 bytesavailable(io::GenericIOBuffer) = io.size - io.ptr + 1
+
+# Position is zero-indexed, but ptr is one-indexed, hence the -1
+# TODO: Document that position for an unseekable stream is invalid, or
+# make it error
```
```suggestion
# TODO: Document that position for an unmarked and unseekable stream is invalid (and make it error?)
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -164,15 +268,43 @@ PipeBuffer(maxsize::Integer) = (x = PipeBuffer(StringMemory(maxsize), maxsize =
+function copy(b::GenericIOBuffer{T}) where T
+    if b.reinit
+        # If buffer is used up, allocate a new size-zero buffer
+        # Reinit implies wriable, and that ptr, size, offset and mark are already the default values
```
```suggestion
        # Reinit implies writable, and that ptr, size, offset and mark are already the default values
```

vtjnash commented on base/iobuffer.jl:
```diff
@@ -1,26 +1,117 @@
+    ) where T<:AbstractVector{UInt8}
+    require_one_based_indexing(data)
+    mz = Int(maxsize)::Int
+    len = Int(length(data))::Int
+    if mz < len
+        throw(ArgumentError("maxsize must not be smaller than data length"))
+    end
```
This error seems a bit arbitrary, since as the comments elsewhere note, the implementation has to support it anyways and it has a fairly well-defined meaning (restricts resizing), and adding the error complicates the constructor situation (with `_new_generic_iobuffer`). Maybe we also instead need to pass the length directly as an argument to this constructor instead of inferring intent from `length(data)`? Anyways, I guess this is just a nit, and not actually important to change.

jakobnissen replied:
The implementation in this PR only supports `length(data) > maxsize` in the case where a `Vector` is passed in, where the legnth of the vector does not exceed `maxsize`, but the underlying memory does. When that is the case, we know that the user never has more than `maxsize` bytes of used data in the buffer. If we ever have that, it becomes a problem, because if we then need to resize (or reallocate the data) for some reason, we discard user data.

vtjnash commented on base/iobuffer.jl:
```diff
@@ -1,26 +1,117 @@
 mutable struct GenericIOBuffer{T<:AbstractVector{UInt8}} <: IO
-    data::T # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
-    reinit::Bool # if true, data needs to be re-allocated (after take!)
+    # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
```
```suggestion
    # T should support: getindex, setindex!, length, copyto!, similar, size, and (optionally) resize!
```

I noticed you added `lastindex` to the list of functions used. That is just a fairly convoluted way of calling `length` though, so I don't really know if that is necessary, but the `require_one_based_indexing` someone previously added already does the same dispatch calls, so it seems okay to just add that to the list here.

jakobnissen committed:
WIP: Address some reviewer comments

jakobnissen committed:
Error when calling `write(x::IOBuffer, x)`

Since it needs to write all its content, this would cause an infinite loop,
exhausing all memory. Throw an error instead.

jakobnissen committed:
Make writing between IOBuffer always exhaust the source

This makes the method write(::IOBuffer, ::IOBuffer) work like the generic
fallback.

jakobnissen committed:
Address some more reviewer comments

jakobnissen committed:
WIP: Address more comments

jakobnissen committed:
Revive dead code in close

jakobnissen committed:
Sink overallocation into existing _resize calls

jakobnissen committed:
Replace zero_offset! with compact!

Which both zeros offset and compacts in one go.

jakobnissen committed:
Do not both compact and allocate new data

If we need to allocate new data, compacting first is a waste of time.

jakobnissen committed:
Fixup

jakobnissen committed:
Revert change in stream

jakobnissen committed:
Track compaction

Switch the offset field to store either an offset, or the number of bytes that
has been compacted. This fixes a current bug where position, mark and reset
did not work correctly with non-seekable buffers.

vtjnash commented on base/iobuffer.jl:
```diff
@@ -322,113 +483,142 @@ function seekend(io::GenericIOBuffer)
+    iszero(offset) && return 0
+    data = io.data
+    copyto!(data, 1, data, offset + 1, length(used_span))
+    io.offset = 0
```
```suggestion
    io.offset -= offset # preserves mark/position computations after compacting shifted the data
    ismarked(io) || (io.offset = 0) # but not required to preserve position/mark computation if the stream was not marked (this step is optional)
```

(plus some changes to `_resize!` / `ensureroom_slowpath` since directly adding `offset + maxsize` is not correct anymore, but rather likely needs to use `min(io.maxsize, length(get_used_span) + n_requested_short)` instead of `min(lastindex(io.data), io.maxsize)`)

jakobnissen replied:
This is handled in [819d750](https://github.com/JuliaLang/julia/pull/57570/commits/819d7506b51d385e534d043042567750687cf621) now.

jakobnissen committed:
Fixup: Revert change in stream

This is done by not erroring with a buffer larger than maxsize if truncate is
set.

jakobnissen commented on base/iobuffer.jl:
```diff
@@ -322,113 +483,142 @@ function seekend(io::GenericIOBuffer)
         end
+        # Max out the buffer size if we want more than the buffer size
+        if length(io.data) < io.maxsize
+            _resize!(io, io.maxsize, true)
```
The offset has been zeroed out in the `compact!` call above, if the offset is nonzero.

jakobnissen committed:
Fixup doctest

jakobnissen committed:
Fixup stream

jakobnissen commented on base/iobuffer.jl:
```diff
@@ -29,14 +104,20 @@ function GenericIOBuffer(data::T, readable::Bool, writable::Bool, seekable::Bool
-    buf.size = length(data) + offset
-    buf.offset = offset
+    if !iszero(offset)
+        unsafe_copyto!(mem, 1, mem, offset+1, len)
```
Going to resolve this, since if `unsafe_copyto!` stops supporting aliasing memory, half of Base will break so this will be the least of our concerns.

jakobnissen committed:
Fixup again

vtjnash commented on base/iobuffer.jl:
```diff
@@ -67,19 +67,26 @@ mutable struct GenericIOBuffer{T<:AbstractVector{UInt8}} <: IO
+    # If compacted: Value is in typemin(Int):0
+    offset_or_compacted::Int
+
+    # The mark is -1 if offset, else the zero-indexed virtual position of ptr in the buffer.
```
```suggestion
    # The mark is -1 if not set, else the zero-indexed virtual position of ptr in the buffer.
```

jakobnissen committed:
Update base/iobuffer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

jakobnissen committed:
Do not recompute available bytes

jakobnissen committed:
Fixup: Reset compacted when taking

jakobnissen committed:
More efficient truncation

jakobnissen committed:
More efficient unmarking

barucden commented on base/iobuffer.jl:
```diff
@@ -1,45 +1,165 @@
+
+    # This field has two distinct meanings:
+    # If the value is positive, it encodes an offset of the start of the data in `data`.
+    # This is used if the buffer is instantiated from a Vector with non-zero memory offet.
```
```suggestion
    # This is used if the buffer is instantiated from a Vector with non-zero memory offset.
```

jakobnissen replied:
Done - I folded this into another commit of mine. :+1:

jakobnissen commented on base/iobuffer.jl:
```diff
@@ -581,31 +831,27 @@ end
-    read_sub(io, b, 1, nr)
-    return nr
+    checkbounds(b, 1:to_read)
+    GC.@preserve b unsafe_read(io, pointer(b), to_read)
```
I'd be willing to do this as a separate PR. :+1: This pattern is used in different ways throughout Base where we want to dispatch on "a chunk of contiguous bytes in CPU memory", it would be nice to coalesce them all to use one mechanism.

jakobnissen committed:
Try allowing writing to maxsize + offset

jakobnissen commented:
Test failures seem unrelated
